### ISSUE TITLE
Unexpected behaviour when defining multiple styles (using Garden units)

### ISSUE THREAD

**Issue by nijk** [Jarzka/stylefy#21]
This may well be an issue caused by a bad implementation, but I'm currently experiencing some expected results when defining multiple styles and components in sequence that use the garden units helpers. It appears that something is causing the style definitions to be reused for some of these heading styles, so instead of styling of `h1-6` I seem to get this:

```
h1: correct
h2: correct
h3: incorrect - uses h2 style & class
h4: incorrect - uses h2 style & class
h5: correct
h6: incorrect - uses h5 style & class
```

Given a list of Reagent component invoking `(use-style ...)` Stylefy appears to re-use the previous style if it is not structurally different from the previous style:

### Styles
```
(ns foo.headings
  (:require [reagent.core :as reagent]
            [stylefy.core :as stylefy]))

(def line-height 1.5)

(def h1-style {:font-size (units/rem 3)
               :line-height line-height
               :padding-bottom (units/rem 2)})

(def h2-style {:font-size (units/rem 2.7)
               :line-height line-height
               :padding-bottom (units/rem 1.7)})

(def h3-style {:font-size (units/rem 2.4)
               :line-height line-height
               :padding-bottom (units/rem 1.4)})

(def h4-style {:font-size (units/rem 2.1)
               :line-height line-height
               :padding-bottom (units/rem 1)})

(def h5-style {:font-size (units/rem 1.9)
               :line-height line-height
               :padding-bottom (units/rem 1)})

(def h6-style {:font-size (units/rem 1.7)
               :line-height line-height
               :padding-bottom (units/rem 1)})

(def p-style {:font-size (units/rem 1.6)
              :line-height line-height
              :padding-bottom (units/rem 1)})
```

### Components
```
(ns foo.text.component
  (:require [stylefy.core :refer [use-style]]
            [foo.headings :refer
             [h1-style h2-style h3-style h4-style h5-style h6-style p-style]]))

(defn h1
  ([children] [:h1 (use-style h1-style) children]))

(defn h2
  ([children] [:h2 (use-style h2-style) children]))

(defn h3
  ([children] [:h3 (use-style h3-style) children]))

(defn h4
  ([children] [:h4 (use-style h4-style) children]))

(defn h5
  ([children] [:h5 (use-style h5-style) children]))

(defn h6
  ([children] [:h6 (use-style h6-style) children]))

(defn p
  ([children] [:blockquote (use-style p-style) children]))
```

### Visual
![screen shot 2018-04-17 at 14 30 34](https://user-images.githubusercontent.com/1615460/38872842-27ec9f2e-424c-11e8-9948-f004bf813518.png)

### Markup
![screen shot 2018-04-17 at 14 30 43](https://user-images.githubusercontent.com/1615460/38872848-2c699f20-424c-11e8-9a69-bebc8f9a12c8.png)

**Comment by Jarzka** [comment]
Thanks for the report once again. :D 

You are correct: If two styles are structurally the same (i.e. the style map hash values are the same), stylefy is going to generate the CSS code once and re-use it. However, it seems that your h3 and h4 styles are different, still they have the same stylefy class, which seems to be incorrect.

Have you tried defining those style maps without Garden, i.e. defining units as strings? I believe this problem is related to using garden units.

**Comment by nijk** [comment]
Haha, yeah sorry to hammer you with issues! 

Yes that all fits with my experience and yes my workaround for now is to define the units as strings.

**Comment by Jarzka** [comment]
This will be fixed in the next release (1.4.1). Style maps with different garden units are now treated structurally different and thus produce different hashes.

**Comment by frankitox** [comment]
Hi @Jarzka. This looks like a proper place to ask you this. I've been looking at the code base and found this:

https://github.com/Jarzka/stylefy/blob/master/src/stylefy/impl/styles.cljs#L40-L43

Is it possible this is a bug? Instead of a when I would write something like this

```
(fn [result prop-key]
  (let [prop-value (prop-key style)]
    (if (utils/is-garden-value? prop-value)
      (assoc result prop-key (compiler/render-css prop-value))
      result)))
```

Let me know and I'll submit a PR if needed.

**Comment by Jarzka** [comment]
I think this is not a bug. The point of this function is to collect all key-value pairs which contain Garden unit values, and convert those values to CSS. The result is stored in `hashable-garden-units`. After that, we merge the result with the original style map: `hashable-style (merge style hashable-garden-units)`

After this, the hashable-style contains the original style map, but with converted Garden values.

**Comment by Jarzka** [comment]
First, I was thinking this is not a bug, but after taking a closer look, I changed my mind.

The function is part of `reduce`. if the `prop-value` is a garden unit, we assoc it to the result. However, if it's not, the function returns nil, which simply destroys all the previous results. This should never happen, and your version should fix it. Thanks for pointing this out. PR would be much welcome.

**Comment by Jarzka** [comment]
I was able to reproduce and fix the bug, so no PR needed. However, I will give you a credit in the changelog for finding this.

I'll release the fix in a couple of days.

**Comment by frankitox** [comment]
That's perfect! Thanks Jari.

**Comment by Jarzka** [comment]
This bug is now fixed in the latest release: 1.14.1
